Reset L1 contracts to pre-MCP

maurelian opened the pull request:
This removes the Multichain changes to the L1 contracts by reverting them to the commit f54a2234f (and then running `forge fmt`). 

This also required reverting the `/universal` dependencies which had been updated (`ERC721Bridge.sol`, `StandardBridge.sol` and `CrossDomainMessenger.sol`). 
And the L2 child contracts (`L2ERC721Bridge.sol`, `L2StandardBridge.sol` and `L2CrossDomainMessenger.sol`) thus also required updating.

Tests and deploy scripts were modified as necessary. 

In order to verify that the layout was properly restored, a pre-mcp storage layout can be found [here](https://github.com/ethereum-optimism/optimism/blob/premcp-storage/packages/contracts-bedrock/.storage-layout#L1). There is no diff whatsoever with the storage layout on this branch.

Fixes https://github.com/ethereum-optimism/client-pod/issues/181

maurelian committed:
feat(ctb): Checked out pre-mcp impl contracts

maurelian committed:
feat(ctb): temporarily restore semver

maurelian committed:
feat(ctb): forge fmt

maurelian committed:
feat(ctb): Update functions which deploy implementations

maurelian committed:
feat(ctb): Update init functions in deploy.sol

maurelian committed:
feat(ctb): Update tests

feat(Update Multichain.s.sol
Update tests

Update L2OutputOracle tests

maurelian committed:
feat(ctb): bindings, semver, snapshot

tynes commented on packages/contracts-bedrock/src/L1/L2OutputOracle.sol:
```diff
@@ -41,14 +38,6 @@ contract L2OutputOracle is Initializable, ISemver {
 
-    /// @notice The address of the challenger. Can be updated via reinitialize.
-    /// @custom:network-specific
-    address public challenger;
```
We should keep the lowercase interfaces with the uppercase ones marked as deprecated but still public

maurelian replied:
Dang, OK will do.

tynes commented on packages/contracts-bedrock/src/universal/Semver.sol:
```diff
@@ -0,0 +1,40 @@
+contract Semver {
+    /// @notice Contract version number (major).
+    uint256 private immutable MAJOR_VERSION;
+
```
I know this is WIP but just saying i'd like to keep this deprecated if possible

maurelian replied:
no cap

maurelian committed:
feat(ctb): Remove Semver again

maurelian committed:
feat(ctb): Remove MCP from OptimismMintableERC20Factory

maurelian committed:
restore lower case getters to tests

maurelian committed:
feat(ctb): Maintain backwards compatibility with lowercase getters

maurelian committed:
fix(ctb): Correctly set values in constructors.

maurelian commented on packages/contracts-bedrock/src/L1/L1CrossDomainMessenger.sol:
```diff
@@ -13,25 +13,24 @@ import { Constants } from "src/libraries/Constants.sol";
     /// @notice Semantic version.
-    /// @custom:semver 1.7.1
-    string public constant version = "1.7.1";
+    /// @custom:semver 1.5.0
+    string public constant version = "1.5.0";
```
```suggestion
    /// @custom:semver 1.8.0
    string public constant version = "1.8.0";
```

maurelian commented on packages/contracts-bedrock/src/L1/L1ERC721Bridge.sol:
```diff
@@ -19,19 +19,13 @@ contract L1ERC721Bridge is ERC721Bridge, ISemver {
     /// @notice Semantic version.
-    /// @custom:semver 1.4.1
-    string public constant version = "1.4.1";
+    /// @custom:semver 1.5.1
+    string public constant version = "1.5.1";
```
```suggestion
    /// @custom:semver 1.5.0
    string public constant version = "1.5.0";
```

maurelian commented on packages/contracts-bedrock/src/L1/L2OutputOracle.sol:
```diff
@@ -15,21 +15,22 @@ contract L2OutputOracle is Initializable, ISemver {
     /// @notice The interval in L2 blocks at which checkpoints must be submitted.
     ///         Although this is immutable, it can safely be modified by upgrading the
     ///         implementation contract.
```
```suggestion
    ///         implementation contract.
    ///         Public getter is legacy and will be removed in the future. Use `submissionInterval`
    ///         instead.
    /// @custom:legacy
```

maurelian commented on packages/contracts-bedrock/src/L1/L2OutputOracle.sol:
```diff
@@ -15,21 +15,22 @@ contract L2OutputOracle is Initializable, ISemver {
-    ///         Public getter is legacy and will be removed in the future. Use
-    //          `finalizationPeriodSeconds` instead.
+    /// @notice The address of the challenger. Can be updated via upgrade. This will be removed in the
+    ///         future, use `proposer` instead.
```
```suggestion
    ///         future, use `challenger` instead.
```

maurelian commented on packages/contracts-bedrock/src/L1/OptimismPortal.sol:
```diff
@@ -99,69 +102,56 @@ contract OptimismPortal is Initializable, ResourceMetering, ISemver {
-        public
-        reinitializer(Constants.INITIALIZER)
-    {
+    /// @param _config Address of the SystemConfig contract.
+    constructor(L2OutputOracle _l2Oracle, address _guardian, bool _paused, SystemConfig _config) {
+        L2_ORACLE = _l2Oracle;
+        GUARDIAN = _guardian;
+        SYSTEM_CONFIG = _config;
```
```suggestion
    /// @param _config Address of the SystemConfig contract.
    constructor(L2OutputOracle _l2Oracle, address _guardian, bool _paused, SystemConfig _systemConfig) {
        L2_ORACLE = _l2Oracle;
        GUARDIAN = _guardian;
        SYSTEM_CONFIG = _systemConfig;
```

maurelian commented on packages/contracts-bedrock/test/invariants/SystemConfig.t.sol:
```diff
@@ -17,7 +17,15 @@ contract SystemConfig_GasLimitLowerBound_Invariant is Test {
+            _batcherHash: bytes32(hex"abcd"), // batcher hash
+            _gasLimit: 30_000_000, // gas limit
+            _unsafeBlockSigner: address(1), // unsafe block signer
+            _config: Constants.DEFAULT_RESOURCE_CONFIG()
```
```suggestion
            _systemConfig: Constants.DEFAULT_RESOURCE_CONFIG()
```

maurelian commented on packages/contracts-bedrock/scripts/Deploy.s.sol:
```diff
@@ -385,11 +387,27 @@ contract Deploy is Deployer {
+            _l2Oracle: l2OutputOracle,
+            _guardian: guardian,
+            _paused: true,
+            _config: systemConfig
```
```suggestion
            _systemConfig: systemConfig
```

maurelian committed:
ctb: Apply suggestions from code review

maurelian commented on packages/contracts-bedrock/.storage-layout:
```diff
@@ -89,19 +80,15 @@
+| Name     | Type                                                             | Slot | Offset | Bytes | Contract                                 |
+|----------|------------------------------------------------------------------|------|--------|-------|------------------------------------------|
+| __gap    | uint256[49]                                                      | 0    | 0      | 1568  | src/L1/L1ERC721Bridge.sol:L1ERC721Bridge |
+| deposits | mapping(address => mapping(address => mapping(uint256 => bool))) | 49   | 0      | 32    | src/L1/L1ERC721Bridge.sol:L1ERC721Bridge |
```
Not really a diff, just a width change.

tynes replied:
I see that deposits is at the same slot so we are good here

maurelian commented on packages/contracts-bedrock/.storage-layout:
You can compare this file with a pre-mcp layout file [here](https://github.com/ethereum-optimism/optimism/blob/premcp-storage/packages/contracts-bedrock/.storage-layout?rgh-link-date=2023-11-08T02%3A20%3A38Z#L1). 
That layout was generated by:
1. checking out f54a2234f (prior to the MCP changes)
2. then adding the missing NFT Bridges to the `storage-snapshot.sh` (c1ba704730dbab8e8686cd9cd2edaa484efe2a40)
3. regenerating the snapshot. 

There is zero diff between these two version of the file, which gives me confidence that the changes were correctly reversed.

maurelian committed:
fix(ctb): Update immutables.go to fix pnpm validate-deploy-config

tynes commented on packages/contracts-bedrock/src/L1/L2OutputOracle.sol:
```diff
@@ -64,86 +60,82 @@ contract L2OutputOracle is Initializable, ISemver {
         FINALIZATION_PERIOD_SECONDS = _finalizationPeriodSeconds;
 
-        initialize({ _startingBlockNumber: 0, _startingTimestamp: 0, _proposer: address(0), _challenger: address(0) });
+        initialize(_startingBlockNumber, _startingTimestamp);
```
nit: keeping kwargs syntax

tynes commented on packages/contracts-bedrock/src/L1/L2OutputOracle.sol:
```diff
@@ -64,86 +60,82 @@ contract L2OutputOracle is Initializable, ISemver {
-        public
-        reinitializer(Constants.INITIALIZER)
-    {
+    function initialize(uint256 _startingBlockNumber, uint256 _startingTimestamp) public initializer {
```
I notice that we are going back from `reinitialize` to `initializer` - this will not work in practice unless during the 2 step upgrade we zero out the `_initialized` storage slot. This will add extra complexity to the upgrade step but will simplify the maintenance of the code. I lean towards adding complexity to the upgrade step so that we can more easily move between contract versions - ie we do not need to `reinitialize(prev value + 1)` on every version of semver bump to ensure that each contract is deployable

maurelian replied:
I agree on that direction.

tynes commented on packages/contracts-bedrock/src/L1/SystemConfig.sol:
```diff
@@ -145,47 +100,25 @@ contract SystemConfig is OwnableUpgradeable, ISemver {
         transferOwnership(_owner);
-
-        // These are set in ascending order of their UpdateTypes.
-        _setBatcherHash(_batcherHash);
```
We should keep the setters here ie emitting the events in the call to `initialize`. It reduces the amount of information the `op-node` needs for genesis

maurelian replied:
OK will do.

maurelian replied:
I did this, and otherwise significantly reduced the diff in SystemConfig. Thanks for calling that out.

tynes commented on packages/contracts-bedrock/src/L1/SystemConfig.sol:
```diff
@@ -203,132 +136,67 @@ contract SystemConfig is OwnableUpgradeable, ISemver {
-            // There is no override and it is not set in storage. Set it to the block number.
-            // This is for newly deployed chains.
-            startBlock = block.number;
+    function unsafeBlockSigner() external view returns (address) {
```
named return arg?

tynes commented on packages/contracts-bedrock/test/FaultDisputeGame.t.sol:
```diff
@@ -139,7 +139,7 @@ contract FaultDisputeGame_Test is FaultDisputeGame_Init {
     ///               For now, it is critical that the first proposed output root of an OP stack
     ///               chain is done so by an honest party.
     function test_initialize_firstOutput_reverts() public {
-        uint256 submissionInterval = l2OutputOracle.submissionInterval();
```
We shouldn't need to make this diff if the new abi is preserved

maurelian replied:
you're correct, but I'm going to keep this diff because it's more consistent with the rest of this test file.

tynes commented on packages/contracts-bedrock/scripts/Deploy.s.sol:
```diff
@@ -816,15 +813,10 @@ contract Deploy is Deployer {
         require(oracle.SUBMISSION_INTERVAL() == cfg.l2OutputOracleSubmissionInterval());
-        require(oracle.submissionInterval() == cfg.l2OutputOracleSubmissionInterval());
         require(oracle.L2_BLOCK_TIME() == cfg.l2BlockTime());
-        require(oracle.l2BlockTime() == cfg.l2BlockTime());
```
Should these getters be deleted?

maurelian replied:
no, I restored them, and reviewed the rest of this file for similar issues.

tynes commented:
We need to update the storage config: https://github.com/ethereum-optimism/optimism/blob/c7f6938e42e7fa6bd7fd5541445392135b3c26c7/op-chain-ops/genesis/config.go#L732

The storage variables that were removed need to be deleted from each contract. `_initialized` and `_initializing`

This might leave an unused variable which if deleted would require a job in CI to be removed

tynes commented:
Looks like this also has a merge conflict now

maurelian committed:
fix(ctb): Restore missing duplicate getters to Deploy script

maurelian committed:
feat(ctb): Named return in systemconfig

maurelian committed:
fix(ctb): Kwargs in l2OO

maurelian committed:
chain-ops: remove initializable from genesis

tynes commented on packages/contracts-bedrock/scripts/ChainAssertions.sol:
```diff
@@ -17,6 +17,7 @@ import { Predeploys } from "src/libraries/Predeploys.sol";
 import { Types } from "scripts/Types.sol";
 import { Vm } from "forge-std/Vm.sol";
 import { ISystemConfigV0 } from "scripts/interfaces/ISystemConfigV0.sol";
+import { console2 as console } from "forge-std/console2.sol";
```
Dead code?

maurelian replied:
No, it's restoring this check: 

https://github.com/ethereum-optimism/optimism/blob/remove-mcp/packages/contracts-bedrock/scripts/ChainAssertions.sol#L132-L135

tynes commented on packages/contracts-bedrock/scripts/Deploy.s.sol:
```diff
@@ -487,30 +512,32 @@ contract Deploy is Deployer {
-        require(config.startBlock() == type(uint256).max);
+        require(resourceConfig.maxResourceLimit == defaultConfig.maxResourceLimit);
+        require(resourceConfig.elasticityMultiplier == defaultConfig.elasticityMultiplier);
+        require(resourceConfig.baseFeeMaxChangeDenominator == defaultConfig.baseFeeMaxChangeDenominator);
```
I thought we were using the modularized library for all of the checks?

maurelian replied:
This is what's on `develop`. It looks like the PR only used the `ChainAssertions` lib in the `initializeX()` functions.

maurelian replied:
After looking into this more, I think it would be best done in a separate PR.

Some of the checks are straightforward, but for other contracts it's fairly involved. 

1. we'll need a new `_implementations()` getter that serves the same purpose as `_proxies()`
2. for contracts where the proxy and implementation have a different set of checks, we'll need two new functions,
  i.e `checkSystemConfig()`, which be split into `checkSystemConfigImpl`, `checkSystemConfigProxy` and `checkSystemConfigCommon`.

Or there might be better ways to do it, but either way it's going to open up new territory for bike shedding.

tynes replied:
Can we open a ticket to capture this work?

maurelian replied:
Sure thing, [219](https://github.com/ethereum-optimism/client-pod/issues/219).

maurelian committed:
fix(chain-mon): Remove mcp changes

maurelian committed:
feat(ctb): Remove Constants.INITIALIZER

maurelian committed:
Remove INITIALIZED value from genesis config

Also disable the ci-match-values-between-files job, but keep its
definition, it looks useful.

maurelian committed:
op-chain-ops: Remove additional SystemConfig interface

maurelian committed:
feat(ctb): Reduce diff in SystemConfig

maurelian committed:
fix(op-chain-ops): Move values out of storage in immutables.go and config.go

maurelian commented:
> We need to update the storage config: https://github.com/ethereum-optimism/optimism/blob/c7f6938e42e7fa6bd7fd5541445392135b3c26c7/op-chain-ops/genesis/config.go#L732

> The storage variables that were removed need to be deleted from each contract. `_initialized` and `_initializing`

> This might leave an unused variable which if deleted would require a job in CI to be removed 

I removed the `check-values-match` [job](https://github.com/ethereum-optimism/optimism/pull/8094/files#diff-78a8a19706dbd2a4425dd72bdab0502ed7a2cef16365ab7030a5a0588927bf47), is that the one?

maurelian committed:
fix(op-chain-ops): Update check-l2

tynes commented on op-chain-ops/genesis/config.go:
```diff
@@ -682,10 +680,14 @@ func NewL2ImmutableConfig(config *DeployConfig, block *types.Block) (immutables.
 	immutable["L2CrossDomainMessenger"] = immutables.ImmutableValues{
 		"otherMessenger": config.L1CrossDomainMessengerProxy,
 	}
+	immutable["OptimismMintableERC20Factory"] = immutables.ImmutableValues{
```
Any reason to move this from the bottom? It just increases the diff

maurelian replied:
The diff at the bottom removes the `storage` entry, this is adding one to `immutable`.

maurelian replied:
I moved this entry to the end of the immutable config just for good measure though.

tynes commented on op-chain-ops/genesis/config.go:
```diff
@@ -26,10 +26,8 @@ import (
-// https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/libraries/Constants.sol
-const InitializedValue = 3
+// InitializedValue represents the `Initializable` contract value. It should always be 1.
+const InitializedValue = 1
```
This is now dead code

maurelian replied:
OK.

maurelian committed:
op-chain-ops: Remove InitializedValue from check-l2
